services/src/main/java/org/keycloak/services/HttpResponseImpl.java

```diff
@@ -44,6 +56,31 @@ public void setHeader(String name, String value) {
         delegate.getOutputHeaders().putSingle(name, value);
     }
 
+    @Override
+    public void addCookieIfAbsent(HttpCookie cookie) {
+        if (cookie == null) {
+            throw new IllegalArgumentException("Cookie is null");
+        }
+
+        if (cookies == null) {
+            cookies = new HashSet<>();
+        }
+
+        if (cookies.add(cookie)) {
+            if (writeCookiesOnTransactionComplete) {
+                // cookies are written after transaction completes
+                return;
+            }
+
+            addHeader(HttpHeaders.SET_COOKIE, cookie.toHeaderValue());
+        }
+    }
+
+    @Override
+    public void setWriteCookiesOnTransactionComplete() {
```

Comment:
Looking at where this is used it looks very weird to me that we'd have oidc authentication endpoint basically wrapped in these retriable transactions, seems like a bit of a hack for some corner-cases, which then has implications on all requests resulting in spinning up new tx/objects/etc, when it's not needed 99.9999% of the time.

Reply:
We can discuss this in more detail in a separate issue but as you pointed out there is an additional overhead to authorization and login action endpoints.

I can create a follow-up issue to discuss it in more detail with the store team.

Reply:
+1 let's do a follow-up around this one

Reply:
These retriable blocks were introduced to deal with CockroachDB's RETRY exceptions that cause a significant amount of requests to fail in concurrent scenarios. Retrying the transaction is their recommended approach in these cases.

Perhaps a better way to handle this would be to surround the original endpoint method in a try-catch block so we can check if the exception being thrown is a retriable exception and only then use the retry mechanism. That is, the original request is handled by the endpoint normally, and we trigger the retry mechanism only if a retriable exception is caught. This would prevent the creation of extra session/tx/objects, which would not be needed in the majority of scenarios.

We could also discuss if this behavior should be configurable at the realm level. Something like "Retry transactions where possible: yes/no", along with maximum number of retries or maximum time allowed for retrying.